core/src/main/java/org/testcontainers/utility/DockerImageName.java

```diff
@@ -125,14 +141,14 @@ public String getUnversionedPart() {
      * @return the versioned part of this name (tag or sha256)
      */
     public String getVersionPart() {
-        return versioning.toString();
+        return versioning == null ? "latest" : versioning.toString();
```

Comment:
can we make `versioning` `@NonNull` and use `Versioning.TagVersioning.LATEST` if `null` is passed to `@Nullable` methods?

Reply:
So, this is one that requires some discussion, and potentially a change or just clear docs!

I'd expect most usage of this feature to be like `foo.asCompatibleSubstituteFor("bar")` meaning `bar` with _any_ tag is accepted.

I wanted to leave the possibility open to specify an exact tag match, i.e. `foo.asCompatibleSubstituteFor("bar:1.2.3")`.

So that this works:
* an absent tag is recorded as `null`
* the compatitility check code treats this `null` as a wildcard
* conversion to a string treats this `null` as an implicit `latest`

It doesn't have to be this way, though. I reckon we could:
* ignore tags altogether for compatibility checks
* OR be more explicit about wildcards, e.g. `foo.asCompatibleSubstituteFor("bar:*")` in the API, and/or have a `Versioning.Wildcard` type internally instead of `null`.

WDYT?

Reply:
> have a `Versioning.Wildcard` type internally instead of `null`

I like this one! My main concern was the Nullable field that we can avoid and I think `Versioning.Wildcard` solves the problem very well 👍

Reply:
👍 cool, I'll go with that then.

Reply:
Done (called it `AnyVersion`)